File: Top200/23-MergeKLists.java
package Top200;

import java.util.Comparator;
import java.util.PriorityQueue;

class MergeKList {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }

        ListNode dummy = new ListNode(-1, null);
        ListNode curr = dummy;

        PriorityQueue<ListNode> queue = new PriorityQueue<ListNode>(lists.length, new Comparator<ListNode>(){
            @Override
            public int compare(ListNode node1, ListNode node2) {
                return node1.val - node2.val;
            }
        });

        for (ListNode head : lists) {
            if (head != null) {
                queue.offer(head);
            }
        }

        while (!queue.isEmpty()) {
            ListNode first = queue.poll();
            curr.next = first;
            curr = curr.next;
            if (curr.next != null) {
                queue.offer(curr.next);
            }
        }

        return dummy.next;
    }
}

File: Top200/23-MergeKLists.go
package main

import "container/heap"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

type ListNodeHeap []*ListNode

func (heap ListNodeHeap) Swap(i, j int) {
	heap[i], heap[j] = heap[j], heap[i]
}

func (heap ListNodeHeap) Less(i, j int) bool {
	return heap[i].Val < heap[j].Val
}

func (heap ListNodeHeap) Len() int {
	return len(heap)
}

func (heap *ListNodeHeap) Push(h interface{}) {
	*heap = append(*heap, h.(*ListNode))
}

func (heap *ListNodeHeap) Pop() interface{} {
	len := heap.Len()
	ret := (*heap)[len-1]
	*heap = (*heap)[:len-1]
	return ret
}

func mergeKLists(lists []*ListNode) *ListNode {
	hp := &ListNodeHeap{}
	heap.Init(hp)

	for _, l := range lists {
		if l != nil {
			heap.Push(hp, l)
		}
	}

	dummy := &ListNode{Val: -1}
	dummy.Next = nil
	curr := dummy

	for hp.Len() != 0 {
		ret := heap.Pop(hp).(*ListNode)
		curr.Next = ret
		curr = curr.Next
		if ret.Next != nil {
			heap.Push(hp, ret.Next)
		}
	}

	return dummy.Next
}
